### PR TITLE
simplify protocol selection

### PR DESCRIPTION
**Pull request by jrudolph**

### CODE REVIEW

File: runtime/src/main/scala/akka/grpc/GrpcProtocol.scala [hunk 1]
```diff
@@ -121,7 +119,13 @@ object GrpcProtocol {
    * Detects which gRPC protocol variant is indicated by a mediatype.
    * @return a [[GrpcProtocol]] matching the request mediatype if known.
    */
-  def detect(mediaType: jmodel.MediaType): Option[GrpcProtocol] = protocols.find(_.mediaTypes.contains(mediaType))
+  def detect(mediaType: jmodel.MediaType): Option[GrpcProtocol] = mediaType.subType match {
+    // FIXME: do we need to check mainType?
+    case "grpc" | "grpc+proto"                   => Some(GrpcProtocolNative)
```
Comment (jrudolph) [inline runtime/src/main/scala/akka/grpc/GrpcProtocol.scala:124]: No need for relatively complex collection logic when we can just match on strings.

File: runtime/src/main/scala/akka/grpc/scaladsl/GrpcMarshalling.scala [hunk 2]
```diff
@@ -45,8 +45,8 @@ object GrpcMarshalling {
 
   def negotiated[T](req: HttpRequest, f: (GrpcProtocolReader, GrpcProtocolWriter) => Future[T]): Option[Future[T]] =
     GrpcProtocol.negotiate(req).map {
-      case (maybeReader, writer) =>
-        maybeReader.map(reader => f(reader, writer)).fold(Future.failed, identity)
+      case (Success(reader), writer) => f(reader, writer)
```
Comment (jrudolph) [inline runtime/src/main/scala/akka/grpc/scaladsl/GrpcMarshalling.scala:48]: Avoid higher order function.

File: runtime/src/main/scala/akka/grpc/GrpcProtocol.scala [hunk 3]
```diff
@@ -121,7 +119,13 @@ object GrpcProtocol {
    * Detects which gRPC protocol variant is indicated by a mediatype.
    * @return a [[GrpcProtocol]] matching the request mediatype if known.
    */
-  def detect(mediaType: jmodel.MediaType): Option[GrpcProtocol] = protocols.find(_.mediaTypes.contains(mediaType))
+  def detect(mediaType: jmodel.MediaType): Option[GrpcProtocol] = mediaType.subType match {
+    // FIXME: do we need to check mainType?
```
Comment (raboof) [inline runtime/src/main/scala/akka/grpc/GrpcProtocol.scala:123]: it's not very likely that we would encounter other types where the subtype exactly matches the constants we know of here, so just matching the subtype might indeed be OK

Comment (jrudolph) [reply runtime/src/main/scala/akka/grpc/GrpcProtocol.scala:123]:
```suggestion
    // mainType is not checked. We assume it's the right one.
```

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**